solutions/azure-hub-spoke-connected-group/bicep/main.bicep

```diff
@@ -0,0 +1,175 @@
+param adminPassword string
+
+
+/*** RESOURCES (HUB) ***/
+
+module hub 'modules/hub.bicep' = {
+  name: 'hub'
+  params: {
+    location: location
+    deployVpnGateway: deployVpnGateway
+    deployAzureBastion: deployAzureBastion
+    deployVirtualMachines: deployVirtualMachines
+  }
+}
+
+/*** RESOURCES (ALL SPOKES) ***/
+
+@description('Next hop to the regional hub\'s Azure Firewall')
+resource routeNextHopToFirewall 'Microsoft.Network/routeTables@2022-01-01' = {
+  name: 'route-to-${location}-hub-fw'
+  location: location
+  properties: {
+    routes: [
+      {
+        name: 'r-nexthop-to-fw'
+        properties: {
+          nextHopType: 'VirtualAppliance'
+          addressPrefix: '0.0.0.0/0'
+          nextHopIpAddress: hub.outputs.firewall.properties.ipConfigurations[0].properties.privateIPAddress
+        }
+      }
+    ]
+  }
+}
+
+/*** RESOURCES (SPOKE ONE) ***/
+module spokenonprod1 'modules/spoke.bicep' = {
+  name: 'spokenonprod1'
+  scope: resourceGroup()
+  params: {
+    location: location
+    deployVirtualMachines: deployVirtualMachines
+    adminUsername: adminUsername
+    adminPassword: adminPassword
+    routeTableId: routeNextHopToFirewall.id
+    spokeName: 'nonprod1'
+    spokeVnetPrefix: '10.100.0.0/22'
+    logAnalyticsWorkspaceId: hub.outputs.logAnalyticsWorkspaceId
+  }
+}
+
+
+/*** RESOURCES (SPOKE TWO) ***/
+module spokenonprod2 'modules/spoke.bicep' = {
+  name: 'spokenonprod2'
+  scope: resourceGroup()
+  params: {
+    location: location
+    deployVirtualMachines: deployVirtualMachines
+    adminUsername: adminUsername
+    adminPassword: adminPassword
+    routeTableId: routeNextHopToFirewall.id
+    spokeName: 'nonprod2'
+    spokeVnetPrefix: '10.101.0.0/22'
+    logAnalyticsWorkspaceId: hub.outputs.logAnalyticsWorkspaceId
+  }
+}
+
+/*** RESOURCES (SPOKE THREE) ***/
+module spokeprod1 'modules/spoke.bicep' = {
+  name: 'spokeprod1'
+  scope: resourceGroup()
+  params: {
+    location: location
+    deployVirtualMachines: deployVirtualMachines
+    adminUsername: adminUsername
+    adminPassword: adminPassword
+    routeTableId: routeNextHopToFirewall.id
+    spokeName: 'prod1'
+    spokeVnetPrefix: '10.200.0.0/22'
+    logAnalyticsWorkspaceId: hub.outputs.logAnalyticsWorkspaceId
+  }
+}
+
+/*** RESOURCES (SPOKE FOUR) ***/
+module spokeprod2 'modules/spoke.bicep' = {
+  name: 'spokeprod2'
+  scope: resourceGroup()
+  params: {
+    location: location
+    deployVirtualMachines: deployVirtualMachines
+    adminUsername: adminUsername
+    adminPassword: adminPassword
+    routeTableId: routeNextHopToFirewall.id
+    spokeName: 'prod2'
+    spokeVnetPrefix: '10.201.0.0/22'
+    logAnalyticsWorkspaceId: hub.outputs.logAnalyticsWorkspaceId
+  }
+}
+
+/*** AZURE VIRTUAL NETWORK MANAGER RESOURCES ***/
+module avnm 'modules/avnm.bicep' = {
+  name: 'avnm'
+  scope: resourceGroup()
+  params: {
+    location: location
+    hubVnetId: hub.outputs.hubVnetId
+    nonProdNetworkGroupMembers: [
```

Comment:
Question on timing here.  I'm kinda surprised to see that all the networks get deployed "naked" (unenrolled) and then, assuming everything goes well, they eventually get connected to AVNM.  Wouldn't the timing be better AVNM was deployed first, if even empty, and then as hubs/spokes showed up they were done so with incremental updates to AVNM so that we have "near realtime" enlistment.  The idea of "unprotected" resources for an indiscriminate amount of time and then play the "catchup process" at the end seems a bit backwards.  Any light you can shed on this?  Is there a technical limitation to delaying adding this management layer until _everything_ is deployed?  If someone wanted to add a third hub or a third spoke, you'd imagine that it would just enlist in the existing AVNM.  As such, it would make sense to do that with the 1st resource as well.

Reply:
Good call out. I think the AVNM answer to this is to use Policy to dynamically associate VNETs with AVNM. I went with static membership for simplicity, and, assuming this is deployed in an environment with other resources, ensuring the test wouldn't interfere with existing VNETs. 

I added this comment on line 23 of azure-hub-spoke-connected-group\bicep\modules\avnm.bicep:
```bicep
      managementGroups: []
    }
  }
  // static network group membership is used to avoid potential conflict in the test environment 
  // for production deployments, consider using Azure Policy to dynamically bring VNETs under 
  // AVNM management. see https://learn.microsoft.com/azure/virtual-network-manager/concept-azure-policy-integration
  resource networkGroupProd 'networkGroups@2022-05-01' = {
    name: 'ng-${location}-spokes-prod'
```

Reply:
I think that comment helps. Thanks.